src/hip/_async/sessions.py

```diff
@@ -0,0 +1,111 @@
+    typing.Mapping[str, str], Cookies,
+]
+DataType = typing.Union[
+    typing.AnyStr,
+    typing.BinaryIO,
+    typing.TextIO,
+    typing.Iterable[typing.AnyStr],
+    # This needs to get unasync-ed into 'typing.Iterable[typing.AnyStr]'
+    typing.AsyncIterable[typing.AnyStr],
+    # This needs to get unasync-ed into 'SyncRequestData'
+    AsyncRequestData,
+]
+JSONType = typing.Union[
+    typing.Mapping["JSONType", "JSONType"],
+    typing.Sequence["JSONType"],
+    int,
+    bool,
+    str,
+    float,
+    None,
+]
+
+
+class Session:
+    """
+    The central instance that manages HTTP life-cycles and interfaces
+    with the background connection pools.
+
+    Adding all of the shortened 'per-method' functions to the
+    Session can be done later once the entire interface is complete.
+    Until that time they are basically just dead-weight for testing
+    and updating.
+    """
+
+    def __init__(self, trust_env: bool = True):
+        self.trust_env = trust_env
+
+    async def request(
+        self,
+        # Request Headers
+        method: str,
+        url: str,
+        headers: HeadersType = None,
+        auth=None,
+        cookies: CookiesType = None,
+        query: QueryType = None,
+        # Request Body
+        data: DataType = None,
+        json: JSONType = None,
+        # Connection
+        timeout=None,
+        proxies=None,
+        http_versions: typing.Sequence[str] = None,
+        # Lifecycle
+        retries=None,
+        redirects: typing.Union[int, bool] = None,
+    ):
+        """Sends a request."""
+        ...
+
+    def prepare_request(
+        self, method: str, url: str, headers=None, auth=None, cookies=None, query=None
+    ) -> Request:
+        """Given all components that contribute to a request sans-body
+        create a Request instance. This method takes all the information from
+        the 'Session' and merges it with info from the .request() call.
+
+        I've renamed 'params' to 'query' because I never understood the name 'params'.
+        Maybe that's just me and we should keep it params?
+
+        The merging that Requests does is essentially: request() overwrites Session
+        level, for 'headers', 'cookies', and 'query' merge the dictionaries and if you
+        receive a value of 'None' for a key at the request() level then you
+        pop that key out of the mapping.
+
+        People seem to understand this merging strategy.
+
+        One edge-case that has been brought up is how to set an 'empty' value
+        versus a 'key without a value' in query string.
+        See this long issue: https://github.com/psf/requests/issues/2651
+
+        - Requests 2 has: {'k': ''} -> '?k=', {'k': None} -> None so there's no way to create '?k'
+        - "Requests 3" has "accepted" this solution: {'k': ''} -> '?k' but then there's suddenly no
+          way to create '?k=' so... doesn't seem like a solution.
+        - I'm proposing we allow [('k',)] -> '?k'?
```

Comment:
Not sure I understand that one as I would expect a dict like `{'k': (,)}` not list. How would your version work? I think using a sentinel value is fine for this edge case, as proposed in the requests issue.

Reply:
I know we discussed this earlier and now you're more in favor of it. Waiting for @njsmith's opinion as well. We're trying to decide between a sentinel or using the monuple approach.

Reply:
Let me summarize the discussion we had. So headers in urllib3 go through `HTTPHeaderDict` which accepts either a dictionary or a list of pairs (and does more complicated things that we're not concerned about here). There are three ways to make `?k` work:

 1. a sentinel value as in `headers={"k": hip.NO_VALUE}` or `headers=[("k", hip.NO_VALUE)]`
 2. accepting string/bytes, as in `headers=b"?k"`
 3. accepting single-value tuples, as in `headers=[("k",)]`

I think we only need one way to make `?k` work. If we decide to accept string/bytes for other reasons then I'd go for option 2, otherwise I'd go for option 3.

Reply:
This is about URL params, not headers, right? header values have to contain at least one non-whitespace character.

Just to check if I remember right: params are similar to headers in that they're *usually* treated as a key-value dict, *but* you can have multiple copies of the same parameter. (Does order matter? For headers, relative order between headers with the same name matters, but order doesn't matter between headers with different names. And, are param keys case-insensitive like headers? Also, header names are ascii-only; are param keys?) And the point here is that params can also have a special "no value" value.

My intuition is that a sentinel makes the most sense, because you can use that with both the dict-like form and the list-like form. I was going to say `None`, but since that means something different in requests, I guess it has to be something like `hip.PARAM_NO_VALUE`.

Reply:
How does the JS `fetch` API handle this case? How does golang handle it?

Reply:
> This is about URL params, not headers, right? header values have to contain at least one non-whitespace character.

Oops, sorry, I got confused here, it's about URL params, I don't think headers have this problem.

So urllib3 just calls [urllib.parse.urlencode](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode), but requests does some [extra processing](https://github.com/psf/requests/blob/v2.22.0/requests/models.py#L82-L107) to suppress keys where the value is None. So eg. `_encode_params({'k1': [1, 2], 'k2': None})` gives `k1=1&k1=2` while `urlencode({'k1': [1, 2], 'k2': None}, doseq=True)` gives `'k1=1&k1=2&k2=None'`.

> Does order matter? For headers, relative order between headers with the same name matters, but order doesn't matter between headers with different names.

RFC 3986 does not seem to define an order. Requests says "Order is retained if data is a list of 2-tuples but arbitrary if parameters are supplied as a dict.". But yes, order is kept when the name is the same: since it can only be done with `urlencode([('k', 1), ('k', 2)])` or `urlencode({'k': [1, 2]}, doseq=True)` and both give `'k=1&k=2'`.

> And, are param keys case-insensitive like headers? Also, header names are ascii-only; are param keys?

I don't think they are case-insensitive or ascii-only since they can be url-encoded, but I need a RFC 3986 expert here to confirm: @sethmlarson. :)

> How does the JS fetch API handle this case?

They just define the query string to be an ASCII-string or null: https://url.spec.whatwg.org/#concept-url-query. So everyone needs to reinvent the wheel here: https://stackoverflow.com/questions/6566456/how-to-serialize-an-object-into-a-list-of-url-query-parameters

> How does golang handle it?

They have `url.Values`: https://golang.org/pkg/net/url/#Values.

>  Values maps a string key to a list of values. It is typically used for query parameters and form values. Unlike in the http.Header map, the keys in a Values map are case-sensitive.

	v := url.Values{}
	v.Set("name", "Ava")
	v.Add("friend", "Jess")
	v.Add("friend", "Sarah")
	v.Add("friend", "Zoe")
	// v.Encode() == "name=Ava&friend=Jess&friend=Sarah&friend=Zoe"

`url.Values` simply maps strings to string lists: https://github.com/golang/go/blob/release-branch.go1.13/src/net/url/url.go#L822-L826 so there's no magic and you can't output `?k` if you want to use that helper.

Reply:
And I'm happy with the sentinel too.

Reply:
It looks like the standard JS API for this is [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams). On parsing, it treats `foo&bar` as equivalent to `foo=&bar=`, and when producing parameter strings it always produces the latter form, with the `=` signs.

E.g. in the firefox console:

```js
» (new URLSearchParams("a=1&a")).getAll("a")
Array [ "1", "" ]
» (new URLSearchParams("a=1&a")).toString()
"a=1&a="
```

This seems to match the URL spec, which says that `foo` and `foo=` are exactly the same: https://url.spec.whatwg.org/#concept-urlencoded-parser

> If *bytes* contains a 0x3D (=), then let *name* be the bytes from the start of *bytes* up to but excluding its first 0x3D (=), and let *value* be the bytes, if any, after the first 0x3D (=) up to the end of *bytes*. If 0x3D (=) is the first byte, then *name* will be the empty byte sequence. If it is the last, then *value* will be the empty byte sequence.
>
> Otherwise, [i.e., if there is no `=`] let *name* have the value of *bytes* and let *value* be the empty byte sequence.

So the behavior of requests/go/JS are all pretty defensible...

Reply:
The other thing to think about is round-tripping: if we want to have a special way to produce a "missing value" here, then people may also expect us to be able to *extract* a missing value, i.e. parse a URL in such a way that `foo` and `foo=` produce different results. But that would violate the URL spec... a parser ought to treat those indistinguishably.

Reply:
re: Ordering, casing, duplicates:
- ordered: Technically should be unordered, but some services depend on order. 
- non-ascii: Allowed with perecent encoding
- duplicates: allowed 
- Case-sensitivity: Pretty sure they're case sensitive.

The round tripping is a sore point for WHATWG URL and one of my few gripes with it. Having an empty string as a value and no value are two semantically different things but WHATWG seem to treat them interchangeably. I don't recommend following the spec perfectly.